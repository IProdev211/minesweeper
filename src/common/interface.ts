export interface GameState {
  boardMap: string[];
  status: number; // 0: lose, 1: win, -1: default
}

export interface ActionType {
  payload: string | number;
}
