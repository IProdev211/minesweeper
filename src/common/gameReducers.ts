import { createSlice } from '@reduxjs/toolkit';
import { GameState, ActionType } from './interface';

function convertMapPayload(payload: string): string[] {
  const rowList: string[] = payload.split('map:')[1].split('\n');
  return rowList.filter((item: string) => !!item.length);
}

const initialState: GameState = {
  boardMap: [],
  status: -1,
};

const gameSlice = createSlice({
  name: 'game',
  initialState,
  reducers: {
    initialiazeGame() {
      return;
    },
    setBoardMap(state: GameState, action: ActionType) {
      state.boardMap = convertMapPayload(action.payload as string);
    },
    setStatus(state: GameState, action: ActionType) {
      const status = action.payload as string;
      state.status = status === '' ? -1 : status === 'OK' ? 1 : 0;
    },
  },
});

export const { initialiazeGame, setBoardMap, setStatus } = gameSlice.actions;

export const gameReducer = gameSlice.reducer;
