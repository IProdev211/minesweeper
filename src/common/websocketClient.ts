export const SOCKET_ENDPOINT = 'wss://hometask.eg1236.com/game1/';

class WebSocketClient {
  private static _socket: WebSocket;

  public static get socket() {
    return this._socket;
  }

  public static set socket(socketConnection: WebSocket) {
    this._socket = socketConnection;
  }

  public static createConnection(url: string = SOCKET_ENDPOINT) {
    if (WebSocketClient.socket) {
      return WebSocketClient.socket;
    }
    const socketConnection = new WebSocket(url);
    WebSocketClient.socket = socketConnection;
    return WebSocketClient.socket;
  }
}

export default WebSocketClient;
