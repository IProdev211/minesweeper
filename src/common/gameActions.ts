import { take, put, call, apply, fork } from 'redux-saga/effects';
import { eventChannel } from 'redux-saga';
import WebSocketClient from './websocketClient';
import { setBoardMap, setStatus } from './gameReducers';

function createSocketChannel(socket: WebSocket) {
  return eventChannel((emit: (input: string | Error) => void) => {
    const handleOnMessage = (event: MessageEvent) => {
      emit(event.data);
    };

    const errorHandler = (errorEvent: any) => {
      emit(new Error(errorEvent.reason || 'Unkown Error'));
    };

    socket.addEventListener('message', handleOnMessage);
    socket.addEventListener('error', errorHandler);

    const unsubscribe = () => {
      socket.removeEventListener('message', handleOnMessage);
    };

    return unsubscribe;
  });
}

function* requestBoardMap(socket: WebSocket) {
  yield apply(socket, socket.send, ['map']);
}

export function* watchOnGame(): any {
  const socket: WebSocket = yield call(WebSocketClient.createConnection);
  const socketChannel = yield call(createSocketChannel, socket);

  while (true) {
    try {
      const data = yield take(socketChannel);
      console.log('data... ', data);
      if (data.includes('map:')) {
        yield put(setBoardMap(data));
      }
      if (data.includes('new:')) {
        yield fork(requestBoardMap, socket);
      }
      if (data.includes('open:')) {
        yield put(setStatus(data.split('open: ')[1]));
        yield fork(requestBoardMap, socket);
      }
    } catch (err) {
      console.error('socket error:', err);
      socketChannel.close();
    }
  }
}
