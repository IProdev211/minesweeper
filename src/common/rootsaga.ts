import { takeLatest } from 'redux-saga/effects';
import { watchOnGame, handleCreateGame } from '../game/gameActions';
import { startGame, initialiazeGame } from '../game/gameReducers';

export function* watcherSaga() {
  yield takeLatest(initialiazeGame.type, watchOnGame);
  yield takeLatest(startGame.type, handleCreateGame);
}
