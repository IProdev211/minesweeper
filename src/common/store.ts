import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';

import { playReducer } from '../play/playReducers';
import { gameReducer } from '../common/gameReducers';
import { optionReducer } from '../option/optionReducers';
import { watcherSaga } from './rootsaga';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    game: gameReducer,
    option: optionReducer,
    play: playReducer,
  },
  middleware: (getDefaultMiddleware) => {
    return [
      ...getDefaultMiddleware({
        thunk: false,
      }),
      sagaMiddleware,
    ];
  },
});

sagaMiddleware.run(watcherSaga);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
