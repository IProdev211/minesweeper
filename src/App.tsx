import React from 'react';

import Board from './play/board';
import * as S from './App.styles';
import Header from './option/header';
import { RootState } from './common/store';
import { useSelector } from 'react-redux';

function App() {
  const { status } = useSelector((state: RootState) => state.game);

  const message = status === 0 ? 'Game Over!' : status === 1 ? 'You Win!' : '';

  return (
    <S.AppContainer>
      <Header />
      <S.BoardContainer>
        <S.Message status={status} data-testid='game-status'>
          {message}
        </S.Message>
        <Board />
      </S.BoardContainer>
    </S.AppContainer>
  );
}

export default App;
