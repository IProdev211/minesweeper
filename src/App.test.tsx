import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';

import App from './App';
import { watcherSaga } from './common/rootsaga';

const sagaMiddleware = createSagaMiddleware();
const mockStore = configureStore([sagaMiddleware]);

const initialState = {
  game: {
    boardMap: [],
  },
  option: {
    level: 0,
  },
  play: {
    flags: [],
  },
};

describe('<App />', () => {
  it('rendered', () => {
    const store = mockStore(initialState);
    const { container } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Select the level - text should be appeared', () => {
    const store = mockStore(initialState);
    const { getByText, getByTestId } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    expect(getByTestId('level-beginner')).toBeInTheDocument();
    expect(getByText('Select the level...')).toBeInTheDocument();
  });
  it('Board should be appeared after selecting the level', async () => {
    sagaMiddleware.run(watcherSaga);
    const store = mockStore({
      ...initialState,
      game: {
        boardMap: [
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
        ],
      },
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    expect(getByTestId('board')).toBeInTheDocument();
  });
  it('Game state should be appeared correctly when game over', async () => {
    sagaMiddleware.run(watcherSaga);
    const store = mockStore({
      ...initialState,
      game: {
        boardMap: [
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□*□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
          '□□□□□□□□□□',
        ],
        status: 0,
      },
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    expect(getByTestId('game-status')).toHaveTextContent('Game Over!');
  });
  it('Game state should be appeared correctly with win', async () => {
    sagaMiddleware.run(watcherSaga);
    const store = mockStore({
      ...initialState,
      game: {
        boardMap: [
          '12121212',
          '12121212',
          '12121212',
          '12121212',
          '12121212',
          '12121212',
          '12121212',
          '12121212',
          '12121212',
          '12121212',
        ],
        status: 1,
      },
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    expect(getByTestId('game-status')).toHaveTextContent('You Win!');
  });
});
