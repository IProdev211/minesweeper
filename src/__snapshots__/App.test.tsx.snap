// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> rendered 1`] = `
<div
  class="MuiBox-root css-1p6th2l"
>
  <div
    class="MuiBox-root css-j7qwjs"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-149xwtl-MuiTypography-root"
    >
      ðŸ’£ MINESWEEPER
    </h4>
    <div
      class="MuiBox-root css-1e9h9dq"
    >
      <div
        class="MuiBox-root css-0"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-1dekchp-MuiButtonBase-root-MuiButton-root"
          data-testid="level-beginner"
          tabindex="0"
          type="button"
        >
          Beginner
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-1dekchp-MuiButtonBase-root-MuiButton-root"
          data-testid="level-intermediate"
          tabindex="0"
          type="button"
        >
          Intermediate
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-1dekchp-MuiButtonBase-root-MuiButton-root"
          data-testid="level-expert"
          tabindex="0"
          type="button"
        >
          Expert
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-o2ixo2-MuiButtonBase-root-MuiButton-root"
        data-testid="restart"
        tabindex="0"
        type="button"
      >
        Restart
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiBox-root css-yaxopm"
  >
    <div
      class="MuiBox-root css-1nygcii"
      data-testid="game-status"
    />
    <h5
      class="MuiTypography-root MuiTypography-h5 css-xr009-MuiTypography-root"
    >
      Select the level...
    </h5>
  </div>
</div>
`;
